core/io/http_session.hxx

```diff
@@ -280,6 +280,15 @@ class http_session : public std::enable_shared_from_this<http_session>
         return stopped_;
     }
 
+    bool idle_timer_expired() const
+    {
+        auto expired = idle_timer_.expiry() <= asio::steady_timer::clock_type::now();
```

Comment:
Are you sure this is sufficient to prevent the race?  I imagine there is a case where this check passes, but the idle-timer still ends up being invoked 'on the next line'.  Rather than relying on timing here, why not mark the session atomically with whether its in-use, or cancel the idle timer after a session is selected, where failure to cancel causes it to retry selecting?

Reply:
Have updated the logic to retry selecting a session if cancelling the idle timer is not successful.